src/v/kafka/server/handlers/fetch.cc

```diff
@@ -1009,20 +1198,36 @@ class fetch_worker {
  */
 class nonpolling_fetch_plan_executor final : public fetch_plan_executor::impl {
 public:
-    explicit nonpolling_fetch_plan_executor(bool debounce = false)
+    nonpolling_fetch_plan_executor()
       : _last_result_size(ss::smp::count, 0)
-      , _debounce(debounce)
       , _fetch_timeout{[this] { _has_progress.signal(); }} {}
 
     /**
      * Executes the supplied `plan` until `octx.should_stop_fetch` returns true.
      */
     ss::future<> execute_plan(op_context& octx, fetch_plan plan) final {
-        if (_debounce) {
+        const auto& fetch_sg
+          = octx.rctx.server().local().fetch_scheduling_group();
+
+        auto fetch_read_strategy
+          = config::shard_local_cfg().fetch_read_strategy();
+        if (
+          fetch_read_strategy
+          == model::fetch_read_strategy::non_polling_with_debounce) {
             co_await ss::sleep(std::min(
               config::shard_local_cfg().fetch_reads_debounce_timeout(),
               octx.request.data.max_wait_ms));
         }
+        // Ensure both fetch debounce and the fetch scheduling group are enabled
+        // before trying to apply any delay.
+        if (
+          fetch_read_strategy
+            == model::fetch_read_strategy::non_polling_with_pid
+          && config::shard_local_cfg().use_fetch_scheduler_group()) {
```

Comment:
Should we warn about this misconfiguration? Maybe on startup?

Reply:
Since the `fetch_read_strategy` can be change dynamically without restarting we should probably warn in the code here rather than at startup.

Reply:
Just warn once though.